### ISSUE TITLE
Implement AR time series Distributions

### ISSUE THREAD

**Issue by lucianopaz** [pymc-devs/pymc4#198]
This is linked to #44. At the moment we don't have any time series distribution. The simplest time series distributions are the [`AR` variants](https://en.wikipedia.org/wiki/Autoregressive_model#Graphs_of_AR(p)_processes). I'm opening this as a smaller (kinda-) begginer friendly issue. It should work on top of [`tfp.sts.Autoregressive`](https://www.tensorflow.org/probability/api_docs/python/tfp/sts/Autoregressive). But eventually, we should wrap most of the distributions in `tfp.sts`.
Here's a [nice blog post](https://blog.tensorflow.org/2019/03/structural-time-series-modeling-in.html) on time series modelling with tfp.

**Comment by Sayam753** [comment]
@lucianopaz I want to work on this issue. Are there any docs for pymc4 to get started?

**Comment by lucianopaz** [comment]
Great @Sayam753! To get started, you should checkout the [basic usage](https://github.com/pymc-devs/pymc4/blob/master/notebooks/basic-usage.ipynb) and [pymc4 design guide](https://github.com/pymc-devs/pymc4/blob/master/notebooks/pymc4_design_guide.ipynb) notebooks. I think that, at least for adding distributions like AR, you shouldn't have go in depth with what goes on with the [core of pymc4](https://github.com/pymc-devs/pymc4/tree/master/pymc4/flow).

**Comment by Sayam753** [comment]
Hi @lucianopaz I tried implementing [tfp.sts.Autoregressive](https://www.tensorflow.org/probability/api_docs/python/tfp/sts/Autoregressive) in the same way the other distributions are written. Here is my implementation in [colab](https://colab.research.google.com/drive/13Mox_tFUy6UpxYZAQjtbSyw9frJUKAv8). I am facing errors in sampling part.  There is no method **sample** in [tfp.sts.Autoregressive](https://www.tensorflow.org/probability/api_docs/python/tfp/sts/Autoregressive) instead it provides similar **prior_sample** method. I tried wrapping the **prior_sample** to **sample** but that also does not work. Can you help figuring out the issue?

**Comment by brianwa84** [comment]
The distribution this goes with is LinearGaussianStateSpaceModel.

See eg
https://colab.research.google.com/github/tensorflow/probability/blob/master/tensorflow_probability/examples/jupyter_notebooks/Structural_Time_Series_Modeling_Case_Studies_Atmospheric_CO2_and_Electricity_Demand.ipynb

**Comment by lucianopaz** [comment]
Thanks @brianwa84! Looking at the source code it seemed like we could implicitly get these by calling the [make_state_space_model](https://github.com/tensorflow/probability/blob/r0.9/tensorflow_probability/python/sts/structural_time_series.py#L134-L159) method. Is that correct? The state space model seems to return the core distribution that does the work for `log_prob` and `sample`.

**Comment by brianwa84** [comment]
Sounds probably correct. One thing to note is that observation noise is
incorporated on the tfp.sts.Sum component, so if you want that, you may
want a tfp.sts.Sum([tfp.sts.Autoregressive(..)])

Brian Patton | Software Engineer | bjp@google.com

**Comment by lucianopaz** [comment]
@Sayam753, i checked out the colab you shared and then looked a bit at the source code in sts. As @brianwa84 pointed out, the autoregressive class isn't a `tfd.Distribution`, we should wrap something different. The structured time series subclasses seem to work more like models that compose many distribution components and provide a nice and clean interface to generate samples from the prior and log probability calculations that are aware of the structural parameters. I'm not completely sure because I didn't have time to read the source code in detail, but it looks like we want to grab the auto regressive class instance, call its `make_state_space_model` to build the underlying `tfd.Distribution`, and ship that out as the protected `_distribution`.

I think it's safe to say that supporting all the things provided by the autoregressive structural model is not beginner friendly. We can split the tasks that should be done into:

1. Manage to get the correct underlying tfd.Distribution.
2. Making the distribution's `sample` method (don't confuse this with `pm.sample`) work.
3. Getting the `log_prob` to work.
4. Add support for missing values in the observations.

Let's just focus on points 1 and 2 for now. Point 4 is completely out of scope, and should be discussed with the rest of the pymc-devs, as we want to have a uniform api that does this for any of our distributions.

**Comment by Sayam753** [comment]
Hi @lucianopaz I tried implementing first three tasks mentioned above. Here is my approach - 
```python
from tensorflow_probability import sts

class AR(sts.Autoregressive):
  r""" 
  Args: same as 
        https://github.com/tensorflow/probability/blob/356cfddef026b3339b8f2a81e600acd2ff8e22b4/tensorflow_probability/python/sts/autoregressive.py#L294:L328
        
        initial_step(Optional) for make_state_space_model function to make LinearGaussianStateSpaceModel Distribution
        https://github.com/tensorflow/probability/blob/356cfddef026b3339b8f2a81e600acd2ff8e22b4/tensorflow_probability/python/sts/structural_time_series.py#L148:L150
  """
  def __init__(self, order, initial_step=0, **kwargs):
    super().__init__(order=order, **kwargs)
    self.order = order
    self.initial_step = initial_step
    self.distribution = self.make_state_space_model(
        num_timesteps=self.order, 
        param_vals={'coefficients': self._coefficients_prior.sample(), 'level_scale': self._level_scale_prior.sample()}, 
        initial_state_prior=self.initial_state_prior, 
        initial_step=self.initial_step)
```
```python
model = AR(50)
sample = model.distribution.sample()
logprob = model.distribution.log_prob(sample)
```
This does seem to work but I have a couple of questions - 
@brianwa84 What is the difference between num_timesteps and order in [AutoregressiveStateSpaceModel](https://www.tensorflow.org/probability/api_docs/python/tfp/sts/AutoregressiveStateSpaceModel)?

@lucianopaz `make_state_space_model` requires coefficients as argument via [param_vals](https://github.com/tensorflow/probability/blob/v0.9.0/tensorflow_probability/python/sts/structural_time_series.py#L143:L144). How to handle coefficients when [coefficients_prior](https://github.com/tensorflow/probability/blob/356cfddef026b3339b8f2a81e600acd2ff8e22b4/tensorflow_probability/python/sts/autoregressive.py#L297:L301) is also given along with [coefficients](https://github.com/tensorflow/probability/blob/356cfddef026b3339b8f2a81e600acd2ff8e22b4/tensorflow_probability/python/sts/autoregressive.py#L144:L147)? This time I explicitly sampled coefficients from [self._coefficients_prior](https://github.com/tensorflow/probability/blob/356cfddef026b3339b8f2a81e600acd2ff8e22b4/tensorflow_probability/python/sts/autoregressive.py#L376).

Can we use `self.distribution` to make this `AR` class compatible with pymc4?

**Comment by Sayam753** [comment]
Building on top of this AR class, I tried to make it compatible with pymc4. 
```python
import sys
import pymc4 as pm
import numpy as np
from tensorflow_probability import sts
from pymc4.distributions.distribution import ContinuousDistribution
sys.argv = sys.argv[:1] # To solve https://github.com/tensorflow/tensorflow/issues/17702

class AR(ContinuousDistribution):

    def __init__(self, name, order, **kwargs):
        super().__init__(name, order=order, **kwargs)

    @staticmethod
    def _init_distribution(conditions: dict):
        order = conditions["order"]
        time_series_object = sts.Autoregressive(order=order)
        distribution = time_series_object.make_state_space_model(
            num_timesteps=order,
            param_vals={'coefficients': time_series_object._coefficients_prior.sample(), 
                        'level_scale': time_series_object._level_scale_prior.sample()},
            initial_state_prior=time_series_object.initial_state_prior,
            initial_step=conditions["initial_step"] if "initial_step" in conditions else 0)
        return distribution

@pm.model
def model():
    m = yield AR('my-regressor', order=2)
    return m

s = model()
trace = pm.sample(s)
```
But i got a `ValueError: No converter defined for TensorListFromTensor`. Here is the traceback - 
```bash
```

**Comment by tirthasheshpatel** [comment]
> ```python
> param_vals={'coefficients': time_series_object._coefficients_prior.sample(), 
>                         'level_scale': time_series_object._level_scale_prior.sample()}
> ```

Did you try:
```python
param_vals={'coefficients': time_series_object._coefficients_prior.sample().numpy().reshape(-1,),
                        'level_scale': time_series_object._level_scale_prior.sample().numpy().reshape(-1,)}
```

**Comment by lucianopaz** [comment]
@Sayam753, sorry that I couldn't reply sooner. First of all, you should open a pull request with what you've done so far. That way we can move our discussion about the implementation specifics over there and leave this issue only to keep track of progress.

But before that, let's go over the questions you made, starting with the parameters of the distribution.

Any time series model will have something like the number of time steps parameter. Let's imagine we have a series of observations for a variable over time. If we wish to model it as a time series distributions we must say how long the time series is. This is done using the `num_timesteps` parameter. I think that we should have it as a mandatory parameter for our `AR` distribution.

After that, the order parameter is something that is characteristic of an auto regressive model (not a property of any time series distribution). The goal of an AR is to try to predict the future outcome of an observation. To do that, it assumes that the _expected value at the next time step_ is a linear combination of whatever was observed a few time steps before. The number of time steps that are looked back in time is controlled by the `order` parameter. If `order=0`, the observation that we expect to see in the immediately next time step is a linear function of the current time step's observation. If `order=1`, the following observation is assumed to be a linear combination of the current time step and one time step to the past. If you add a bigger order, more time steps of history are assumed to determine the future expected value.

The `coefficients` parameter controls the weight of the previous time steps in the linear combination I was talking about before. The level scale controls the variance that we assume each previous time step adds to the future observation. We should accept tensor values for them, not distributions for which you call `sample` on, like what you're doing in the code you pasted above. What I mean by this is that we should completely disallow the usage of `coefficients_prior` and `level_scale_prior`. I imagine that the reason that tfp adds these is because sts is intended to be used almost in isolation, and it should be possible to specify some hierarchical Bayesian model that let's one infer global parameters from an observed time series. We take a different approach, the time series will be simple distributions, and the hierarchical dependence between parameters will be declared within a `pymc4.model`. I'm sure that the tfp folks would be able to do something similar with their joint distributions and I can only imagine that they wrote sts this way because they expect users will use it directly as simpler high level interface.

Well that's mostly it about the parameter stuff. We haven't addressed the initial state prior stuff but we could do this later, with a similar philosophy as above (disallow prior distributions, and only let users pass tensors in).

Finally, do not attempt to `pm.sample` with xla set to true yet. It's still a tfp experimental feature. The particular error you are seeing seems to indicate that you passed in an integer instead of a float. I've found that tensorflow is very strict with this, even when it shouldn't. I imagine that xla might be complaining because you passed `order=2` and not `oder=2.` (or maybe something similar with the initial step). I wouldn't worry about that yet. Just focus on being able to evaluate an observation's log_prob and to get the AR distribution to work with the `pm.evaluate_model` function.

**Comment by Sayam753** [comment]
Thanks @lucianopaz for the detailed explanation. I will open a pull request soon by incorporating all the changes.

**Comment by lucianopaz** [comment]
Closed by #220
